cpu/sam0_common/periph/spi.c

```diff
@@ -164,28 +203,114 @@ void spi_release(spi_t bus)
     mutex_unlock(&locks[bus]);
 }
 
-void spi_transfer_bytes(spi_t bus, spi_cs_t cs, bool cont,
-                        const void *out, void *in, size_t len)
+static void _blocking_transfer(spi_t bus, const void *out, void *in, size_t len)
 {
     const uint8_t *out_buf = out;
     uint8_t *in_buf = in;
 
-    assert(out || in);
-
-    if (cs != SPI_CS_UNDEF) {
-        gpio_clear((gpio_t)cs);
-    }
-
     for (int i = 0; i < (int)len; i++) {
         uint8_t tmp = (out_buf) ? out_buf[i] : 0;
-        while (!(dev(bus)->INTFLAG.reg & SERCOM_SPI_INTFLAG_DRE)) {}
         dev(bus)->DATA.reg = tmp;
         while (!(dev(bus)->INTFLAG.reg & SERCOM_SPI_INTFLAG_RXC)) {}
         tmp = (uint8_t)dev(bus)->DATA.reg;
         if (in_buf) {
             in_buf[i] = tmp;
         }
     }
+}
+
+#ifdef MODULE_PERIPH_DMA
+
+static void _dma_launch(spi_t bus)
+{
+#if defined(CPU_FAM_SAMD21)
+    pm_block(SAMD21_PM_IDLE_1);
+#endif
+    dma_start(_dma_state[bus].rx_dma);
+    dma_start(_dma_state[bus].tx_dma);
+
+    dma_wait(_dma_state[bus].rx_dma);
+#if defined(CPU_FAM_SAMD21)
+    pm_unblock(SAMD21_PM_IDLE_1);
+#endif
+}
+
+static void _dma_transfer(spi_t bus, uint8_t *out, uint8_t *in, size_t len)
+{
+    uint8_t tmp = 0;
+    uint8_t *out_addr = out ? out + len : &tmp;
+    uint8_t *in_addr = in ? in + len : &tmp;
+    dma_prepare_dst(_dma_state[bus].rx_dma, in_addr, len, in ? true : false);
+    dma_prepare_src(_dma_state[bus].tx_dma, (void*)out_addr, len,
+                    out ? true : false);
+    _dma_launch(bus);
+}
+
+static void _dma_transfer_regs(spi_t bus, uint8_t reg, uint8_t *out,
+                               uint8_t *in, size_t len)
+{
+    uint8_t tmp;
+    uint8_t *out_addr = out ? out + len : &tmp;
+    uint8_t *in_addr = in ? in + len : &tmp;
+
+    dma_prepare_dst(_dma_state[bus].rx_dma, &tmp, 1, false);
+    dma_prepare_src(_dma_state[bus].tx_dma, &reg, 1, false);
+
+    dma_append_dst(_dma_state[bus].rx_dma, &rx_desc[bus], in_addr,
+                   len, in ? true : false);
+    dma_append_src(_dma_state[bus].tx_dma, &tx_desc[bus], out_addr,
+                   len, out ? true : false);
+
+    _dma_launch(bus);
+}
+void spi_transfer_regs(spi_t bus, spi_cs_t cs,
+                       uint8_t reg, const void *out, void *in, size_t len)
+{
+    if (cs != SPI_CS_UNDEF) {
+        gpio_clear((gpio_t)cs);
+    }
+
+    if (_use_dma(bus)) {
+        /* The DMA promises not to modify the const out data */
```

Comment:
Why not *actually* promise it and make the in `src` parameter `const`

Reply:
Done in #14279